### PR TITLE
Lambda deployment type

### PR DESCRIPTION
**Pull request by gidsg**
Adds a deployment type `aws-lambda` which provides one deploy action, `updateLambda`, that runs Lambda Update Function Code using the package files which can be specified per stage.

CC @guardian/discussion @philwills

### CODE REVIEW

**Comment by kenoir** [comment]
Nice!

**Comment by akash1810** [comment]
Nice! Was just thinking how we'd deploy lambdas on the Grid too!

File: magenta-lib/src/main/scala/magenta/deployment_type/Lambda.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      """.stripMargin
+  ).default(Map.empty)
+
+  def perAppActions = {
+    case "updateLambda" => (pkg) => (resourceLookup, parameters, stack) => {
+      implicit val keyRing = resourceLookup.keyRing(parameters.stage, pkg.apps.toSet, stack)
+      val fNames = functionNames(pkg)
+      val stage = parameters.stage.name
+      val fName = fNames get stage
+      assert(fName.isDefined, s"functionName must be defined for stage $stage")
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/deployment_type/Lambda.scala]: The normal pattern in RiffRaff is to use `MessageBroker.fail`, e.g. https://github.com/guardian/deploy/blob/master/magenta-lib/src/main/scala/magenta/tasks/AWS.scala#L122

File: magenta-lib/src/main/scala/magenta/tasks/AWS.scala [hunk 2]
```diff
@@ -7,17 +7,55 @@ import com.amazonaws.services.ec2.AmazonEC2Client
...
 
 trait S3 extends AWS {
   def s3client(keyRing: KeyRing) = new AmazonS3Client(credentials(keyRing))
 }
 
+trait Lambda extends AWS {
+
+  def lambdaClient(implicit keyRing: KeyRing) = {
+    val client = new AWSLambdaClient(credentials(keyRing))
+    client.setEndpoint("lambda.eu-west-1.amazonaws.com")
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/tasks/AWS.scala]: Eek, we should stop hardcoding these at some point, but it's fine for the purpose of this PR.

File: magenta-lib/src/main/scala/magenta/tasks/AWS.scala [hunk 3]
```diff
@@ -7,17 +7,55 @@ import com.amazonaws.services.ec2.AmazonEC2Client
...
+    val request = new UpdateFunctionCodeRequest
+    request.withFunctionName(functionName)
+    request.withZipFile(buffer)
+    request
+  }
+}
+
+object Lambda {
+  def lambdaUpdateFunctionCodeRequest(functionName: String, file: File): UpdateFunctionCodeRequest = {
+
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/tasks/AWS.scala]: How come we have this function in the trait and the object?

**Comment by philwills** [comment]
Apart from those couple of minor points, this looks good.

**Comment by janua** [comment]
@gidsg How does this work if you have a standalone project?

**Comment by gidsg** [comment]
I talked with @janua IRL but please see example project https://github.com/guardian/discussion-platform/tree/riff-raff-lambda/cachebust

File: magenta-lib/src/main/scala/magenta/deployment_type/Lambda.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    List(UpdateLambda(new File(pkg.srcDir.getPath + "/lambda.zip"), fName.get))
```
Comment (janua) [inline magenta-lib/src/main/scala/magenta/deployment_type/Lambda.scala]: @gidsg Would it be possible to put this into a parameter?

Comment (gidsg) [inline magenta-lib/src/main/scala/magenta/deployment_type/Lambda.scala]: What is the benefit of this v.s. extra config? A bit easier build script?

Comment (janua) [inline magenta-lib/src/main/scala/magenta/deployment_type/Lambda.scala]: More explicit instead of an arbitrary filename.

Comment (philwills) [inline magenta-lib/src/main/scala/magenta/deployment_type/Lambda.scala]: RiffRaff params can have defaults, so we can have some flexibility without demanding extra config.

**Comment by gidsg** [comment]
@philwills I made the changes you suggested.

In addition, I have done a refactor with @janua which allows for the following:
* Optionally specify Lambda filenames
* Allow for different artefacts per stage. e.g. for @janua's use case of injecting different config per stage.

Please take a look and let me know if you are OK with these changes.

**Comment by janua** [comment]
Yes!! :+1:

**Comment by philwills** [comment]
:+1: great work @gidsg

**Pull request merged**